src/proxy/pool.rs

```diff
@@ -12,80 +12,539 @@
+        // or they have, but no conns are currently available
+        // (because someone else has checked all of them out and not put any back yet)
+        //
+        // So, we will take a nonexclusive readlock on the lockmap, to see if an inner lock
+        // exists for our key.
+        //
+        // If not, we insert one.
+        {
+            debug!(
+                "didn't find a connection for key {:#?}, making sure lockmap has entry",
+                hash_key
+            );
+            let guard = self.state.established_conn_writelock.guard();
+            match self.state.established_conn_writelock.try_insert(
+                hash_key,
+                Some(Arc::new(Mutex::new(()))),
+                &guard,
+            ) {
+                Ok(_) => {
+                    debug!("inserting conn mutex for key {:#?} into lockmap", hash_key);
+                }
+                Err(_) => {
+                    debug!("already have conn for key {:#?} in lockmap", hash_key);
+                }
+            }
+        }
+
+        // If we get here, it means the following are true:
+        // 1. We have a guaranteed sharded mutex in the outer map for our current key.
+        // 2. We can now, under readlock(nonexclusive) in the outer map, attempt to
+        // take the inner writelock for our specific key (exclusive).
+        //
+        // This doesn't block other tasks spawning connections against other keys, but DOES block other
+        // tasks spawning connections against THIS key - which is what we want.
+
+        // NOTE: This inner, key-specific mutex is a tokio::async::Mutex, and not a stdlib sync mutex.
+        // these differ from the stdlib sync mutex in that they are (slightly) slower
+        // (they effectively sleep the current task) and they can be held over an await.
+        // The tokio docs (rightly) advise you to not use these,
+        // because holding a lock over an await is a great way to create deadlocks if the await you
+        // hold it over does not resolve.
+        //
+        // HOWEVER. Here we know this connection will either establish or timeout
+        // and we WANT other tasks to go back to sleep if a task is already trying to create a new connection for this key.
+        //
+        // So the downsides are actually useful (we WANT task contention -
+        // to block other parallel tasks from trying to spawn a connection for this key if we are already doing so)
+        let inner_conn_lock = {
+            trace!("fallback attempt - getting keyed lock out of lockmap");
+            let guard = self.state.established_conn_writelock.guard();
+
+            let exist_conn_lock = self
+                .state
+                .established_conn_writelock
+                .get(&hash_key, &guard)
+                .unwrap();
+            trace!("fallback attempt - got keyed lock out of lockmap");
+            exist_conn_lock.as_ref().unwrap().clone()
+        };
```

Comment:
If the mutex contained a canonical WorkloadKey that the guard is for we could detect here that a collision had occurred, which I think would be safer than allowing any pool manipulation under unchecked collision.

Reply:
It's part of the ConnClient, which is as close to the horse's mouth as you can get. And where we get a lock, we get a ConnClient, so I think it will be the same, but I'll check if it simplifies the validation.


There would be no difference in safety. Popping a conn is not harmful either way, as long as we do not return it to the caller if there's a mismatch - collision doesn't matter.